File: src/utils/BehaviorTracker.ts
/**
 * BehaviorTracker.ts
 * Captures user interaction patterns to infer personality traits.
 */

export interface InteractionMetrics {
    lastMoveTime: number;
    mousePathLength: number;
    clickCount: number;
    startTime: number;
    stepStartTime: number;
    decisionLatencies: number[];
    velocityPeaks: number[];
    revisionCount: number;
    jitterSum: number;
    contextualChoices: {
        aesthetics?: string;
        primaryTrait?: string;
        traitWeights?: Record<string, number>;
    };
}

export class BehaviorTracker {
    private metrics: InteractionMetrics = {
        lastMoveTime: Date.now(),
        mousePathLength: 0,
        clickCount: 0,
        startTime: Date.now(),
        stepStartTime: Date.now(),
        decisionLatencies: [],
        velocityPeaks: [],
        revisionCount: 0,
        jitterSum: 0,
        contextualChoices: {}
    };

    private lastPos: { x: number; y: number } | null = null;
    private firstPos: { x: number; y: number } | null = null;

    /**
     * Start/Reset tracking for a specific session
     */
    public startSession() {
        this.metrics = {
            lastMoveTime: Date.now(),
            mousePathLength: 0,
            clickCount: 0,
            startTime: Date.now(),
            stepStartTime: Date.now(),
            decisionLatencies: [],
            velocityPeaks: [],
            revisionCount: 0,
            jitterSum: 0,
            contextualChoices: {}
        };
        this.lastPos = null;
        this.firstPos = null;
    }

    /**
     * Call this when a significant decision step is completed
     */
    public recordStepCompletion() {
        const now = Date.now();
        const latency = now - this.metrics.stepStartTime;
        this.metrics.decisionLatencies.push(latency);
        this.metrics.stepStartTime = now;
    }

    /**
     * Backward-compatible alias for tests and older UI integrations.
     */
    public recordDecision() {
        this.recordStepCompletion();
    }

    /**
     * Call this whenever a user changes a previously made choice
     */
    public recordRevision() {
        this.metrics.revisionCount++;
    }

    /**
     * Record a specific contextual choice
     */
    public recordChoice(key: keyof InteractionMetrics['contextualChoices'], value: unknown) {
        this.metrics.contextualChoices[key] = value;
    }

    /**
     * Hook into mouse movement
     */
    public trackMovement(x: number, y: number) {
        const now = Date.now();
        if (!this.firstPos) {
            this.firstPos = { x, y };
        }

        if (this.lastPos) {
            const dx = x - this.lastPos.x;
            const dy = y - this.lastPos.y;
            const dist = Math.sqrt(dx * dx + dy * dy);
            const dt = now - this.metrics.lastMoveTime || 1;

            this.metrics.mousePathLength += dist;

            // Velocity peak tracking
            const velocity = dist / dt;
            if (velocity > 5) { // Threshold for "fast" movement
                this.metrics.velocityPeaks.push(velocity);
            }

            // Jitter/Directness tracking (Sudden direction changes)
            // Simplified: adding up movements that are very small but frequent
            if (dist > 0 && dist < 10) {
                this.metrics.jitterSum += 1;
            }
        }
        this.lastPos = { x, y };
        this.metrics.lastMoveTime = now;
    }

    /**
     * Backward-compatible alias for tests and older UI integrations.
     */
    public recordMouseMove(x: number, y: number) {
        this.trackMovement(x, y);
    }

    public recordClick() {
        this.metrics.clickCount++;
    }

    /**
     * Summarize data for backend consumption
     */
    public getBehavioralProfile() {
        const avgLatency = this.metrics.decisionLatencies.length > 0
            ? this.metrics.decisionLatencies.reduce((a, b) => a + b, 0) / this.metrics.decisionLatencies.length
            : 0;
        const directDistance = this.firstPos && this.lastPos
            ? Math.sqrt(
                (this.lastPos.x - this.firstPos.x) ** 2 +
                (this.lastPos.y - this.firstPos.y) ** 2
            )
            : 0;
        const pathEfficiency = this.metrics.mousePathLength > 0
            ? Math.min(1, directDistance / this.metrics.mousePathLength)
            : 0;

        return {
            pathEfficiency,
            avgDecisionLatency: avgLatency,
            revisionRate: this.metrics.revisionCount,
            jitterIndex: this.metrics.jitterSum / (this.metrics.mousePathLength || 1),
            intensity: this.metrics.velocityPeaks.length > 0
                ? Math.max(...this.metrics.velocityPeaks)
                : 0,
            contextualChoices: this.metrics.contextualChoices
        };
    }
}

export const tracker = new BehaviorTracker();

File: UnrealEngine/BehaviorTracker.cpp
// BehaviorTracker.cpp
// 언리얼 엔진용 행동 수집 및 평가 시스템 구현

#include "BehaviorTracker.h"
#include "Engine/Engine.h"
#include "HAL/PlatformTime.h"

UBehaviorTracker::UBehaviorTracker(const FObjectInitializer& ObjectInitializer)
    : Super(ObjectInitializer)
    , FastMovementThreshold(5.0f)
    , JitterThreshold(10.0f)
{
    StartSession();
}

void UBehaviorTracker::StartSession()
{
    Metrics = FInteractionMetrics();
    Metrics.SessionStartTime = GetCurrentTime();
    Metrics.StepStartTime = GetCurrentTime();
    Metrics.LastMoveTime = GetCurrentTime();
    LastPosition = FVector2D::ZeroVector;
}

void UBehaviorTracker::TrackMovement(const FVector2D& MousePosition)
{
    const float CurrentTime = GetCurrentTime();
    
    if (LastPosition != FVector2D::ZeroVector)
    {
        // 거리 계산
        const float Distance = CalculateDistance(LastPosition, MousePosition);
        const float DeltaTime = CurrentTime - Metrics.LastMoveTime;
        
        if (DeltaTime > 0.0f)
        {
            // 경로 길이 누적
            Metrics.MousePathLength += Distance;
            
            // 속도 계산 (픽셀/ms)
            const float Velocity = Distance / DeltaTime;
            
            // 빠른 움직임 감지
            if (Velocity > FastMovementThreshold)
            {
                Metrics.VelocityPeaks.Add(Velocity);
            }
            
            // 지터 감지 (작고 빈번한 움직임)
            if (Distance > 0.0f && Distance < JitterThreshold)
            {
                Metrics.JitterSum += 1.0f;
            }
        }
    }
    
    LastPosition = MousePosition;
    Metrics.LastMousePosition = MousePosition;
    Metrics.LastMoveTime = CurrentTime;
}

void UBehaviorTracker::RecordClick()
{
    Metrics.ClickCount++;
}

void UBehaviorTracker::StartDecisionStep()
{
    Metrics.StepStartTime = GetCurrentTime();
}

void UBehaviorTracker::RecordStepCompletion()
{
    const float CurrentTime = GetCurrentTime();
    const float Latency = CurrentTime - Metrics.StepStartTime;
    
    if (Latency > 0.0f)
    {
        Metrics.DecisionLatencies.Add(Latency);
    }
    
    // 다음 단계를 위한 초기화
    Metrics.StepStartTime = CurrentTime;
}

void UBehaviorTracker::RecordRevision()
{
    Metrics.RevisionCount++;
}

void UBehaviorTracker::RecordChoice(const FString& Key, const FString& Value)
{
    Metrics.ContextualChoices.Add(Key, Value);
}

FBehavioralProfile UBehaviorTracker::GetBehavioralProfile() const
{
    FBehavioralProfile Profile;
    
    const float TotalTime = GetCurrentTime() - Metrics.SessionStartTime;
    
    // 경로 효율성 계산
    // 실제 경로 길이를 시간으로 나눈 값 (픽셀/ms)
    if (TotalTime > 0.0f)
    {
        Profile.PathEfficiency = Metrics.MousePathLength / TotalTime;
    }
    
    // 평균 의사결정 지연시간
    if (Metrics.DecisionLatencies.Num() > 0)
    {
        float Sum = 0.0f;
        for (float Latency : Metrics.DecisionLatencies)
        {
            Sum += Latency;
        }
        Profile.AvgDecisionLatency = Sum / Metrics.DecisionLatencies.Num();
    }
    
    // 수정 빈도
    Profile.RevisionRate = Metrics.RevisionCount;
    
    // 지터 지수 (지터 합 / 경로 길이)
    if (Metrics.MousePathLength > 0.0f)
    {
        Profile.JitterIndex = Metrics.JitterSum / Metrics.MousePathLength;
    }
    
    // 상호작용 강도 (최대 속도 피크)
    if (Metrics.VelocityPeaks.Num() > 0)
    {
        float MaxVelocity = 0.0f;
        for (float Velocity : Metrics.VelocityPeaks)
        {
            if (Velocity > MaxVelocity)
            {
                MaxVelocity = Velocity;
            }
        }
        Profile.Intensity = MaxVelocity;
    }
    
    // 컨텍스트 선택 복사
    Profile.ContextualChoices = Metrics.ContextualChoices;
    
    return Profile;
}

float UBehaviorTracker::CalculateDistance(const FVector2D& Pos1, const FVector2D& Pos2) const
{
    const float DX = Pos2.X - Pos1.X;
    const float DY = Pos2.Y - Pos1.Y;
    return FMath::Sqrt(DX * DX + DY * DY);
}

float UBehaviorTracker::GetCurrentTime() const
{
    // 밀리초 단위로 시간 반환
    return static_cast<float>(FPlatformTime::Seconds() * 1000.0);
}
